# Tensor Assingment

This is the PyTorch tensor assingment for the leaky.ai Introduction to Deep Learning with PyTorch course (www.leaky.ai).  

To execute this assingment head over to Google Colab and load the notebook.

DIRECTIONS
1.  Load notebook into colab.research.google.com
2.  Replace the [TBD]'s with your own code
3.  Execute the notebook after completing each cell and check your answers

In [1]:
# Import PyTorch and check the version
import torch
torch.manual_seed(6)
torch.__version__

'1.9.0'

### Correct Answer:  
<pre>
'1.9.0'</pre>

# Creating Tensors
Lets create a few tensors that we will use for the remaining part of the exercise

In [2]:
# Create a PyTorch tensor with the following content
# Tensor a should contain [[1,2,3,4,5],[6,7,8,9,10],[11,12,13,14,15],[16,17,18,19,20]]
a = torch.tensor([[1,2,3,4,5],[6,7,8,9,10],[11,12,13,14,15],[16,17,18,19,20]], dtype=torch.float)
print (a)

tensor([[ 1.,  2.,  3.,  4.,  5.],
        [ 6.,  7.,  8.,  9., 10.],
        [11., 12., 13., 14., 15.],
        [16., 17., 18., 19., 20.]])


### Correct Answer:  
<pre>
tensor([[ 1.,  2.,  3.,  4.,  5.],
        [ 6.,  7.,  8.,  9., 10.],
        [11., 12., 13., 14., 15.],
        [16., 17., 18., 19., 20.]])
</pre>

# Exploring Tensor Properties
Here you will explore different properties of tensors including type and sizing.

In [3]:
# Check the size of the tensor
a.size()

torch.Size([4, 5])

### Correct Answer:  
<pre>
torch.Size([4, 5])
</pre>

In [4]:
# Check the type
a.type()

'torch.FloatTensor'

### Correct Answer:  
<pre>
'torch.FloatTensor'
</pre>

# Indexing and Slicing Tensors
Here you will explore indexing and slicing of PyTorch tensors.

In [5]:
# Select the middle values for all the rows
a[:,2]

tensor([ 3.,  8., 13., 18.])

### Correct Answer:  
<pre>
tensor([ 3.,  8., 13., 18.])
</pre>

In [6]:
# Select the last value in row 2 
a[2,-1]

tensor(15.)

### Correct Answer:  
<pre>
tensor(15.)
</pre>

In [7]:
# Print the value 12 as a python scalar
a[2,1].item()

12.0

### Correct Answer:  
<pre>
12.0
</pre>

In [8]:
# Print out 7, 8, 9
a[1,1:4]

tensor([7., 8., 9.])

### Correct Answer:  
<pre>
tensor([7., 8., 9.])
</pre>

In [9]:
# Print out 9, 14, 19 as a Python list
a[1:4,3].tolist()

[9.0, 14.0, 19.0]

### Correct Answer:  
<pre>
[9.0, 14.0, 19.0]
</pre>

In [10]:
# Create a new tensor b by starting with a, 
# subtracting 10 and then multiplying by 2
b = (a-10)*2
print (b)

tensor([[-18., -16., -14., -12., -10.],
        [ -8.,  -6.,  -4.,  -2.,   0.],
        [  2.,   4.,   6.,   8.,  10.],
        [ 12.,  14.,  16.,  18.,  20.]])


### Correct Answer:  
<pre>
tensor([[-18., -16., -14., -12., -10.],
        [ -8.,  -6.,  -4.,  -2.,   0.],
        [  2.,   4.,   6.,   8.,  10.],
        [ 12.,  14.,  16.,  18.,  20.]])
</pre>

# Finding Maximum Values
Here you will explore different approachs to extract the highest values from tensors.

In [11]:
# Find the highest and lowest value in the entire tensor 
maximum = a.max()
minimum = a.min()

print (f"Max: {maximum.item()}\nMin: {minimum.item()}")

Max: 20.0
Min: 1.0


In [12]:
# Find the top 2 maximum values and index in each
val, idx = a.topk(2)
print (f"Values:\n{val}")
print (f"\nIndicies:\n{idx}")

Values:
tensor([[ 5.,  4.],
        [10.,  9.],
        [15., 14.],
        [20., 19.]])

Indicies:
tensor([[4, 3],
        [4, 3],
        [4, 3],
        [4, 3]])


### Correct Answer:  
<pre>
Values:
tensor([[ 5.,  4.],
        [10.,  9.],
        [15., 14.],
        [20., 19.]])

Indicies:
tensor([[4, 3],
        [4, 3],
        [4, 3],
        [4, 3]])
</pre>

# Standardize Columns
For tabular datasets, you will likely need to standardize or normalize the data before using it for training.

In [13]:
# Standardize the column values
mean = a.mean(dim=0, keepdim=True)
std = a.std(dim=0, keepdim=True)
a_norm = (a - mean) / std

# Output the results
print (a)
print (f"\nMeans:\n{mean}")
print (f"\nStandard Deviations:\n{std}")
print (f"\nStandardized:\n{a_norm}")

tensor([[ 1.,  2.,  3.,  4.,  5.],
        [ 6.,  7.,  8.,  9., 10.],
        [11., 12., 13., 14., 15.],
        [16., 17., 18., 19., 20.]])

Means:
tensor([[ 8.5000,  9.5000, 10.5000, 11.5000, 12.5000]])

Standard Deviations:
tensor([[6.4550, 6.4550, 6.4550, 6.4550, 6.4550]])

Standardized:
tensor([[-1.1619, -1.1619, -1.1619, -1.1619, -1.1619],
        [-0.3873, -0.3873, -0.3873, -0.3873, -0.3873],
        [ 0.3873,  0.3873,  0.3873,  0.3873,  0.3873],
        [ 1.1619,  1.1619,  1.1619,  1.1619,  1.1619]])


### Correct Answer:  
<pre>
tensor([[ 1.,  2.,  3.,  4.,  5.],
        [ 6.,  7.,  8.,  9., 10.],
        [11., 12., 13., 14., 15.],
        [16., 17., 18., 19., 20.]])

Means:
tensor([[ 8.5000,  9.5000, 10.5000, 11.5000, 12.5000]])

Standard Deviations:
tensor([[6.4550, 6.4550, 6.4550, 6.4550, 6.4550]])

Standardized:
tensor([[-1.1619, -1.1619, -1.1619, -1.1619, -1.1619],
        [-0.3873, -0.3873, -0.3873, -0.3873, -0.3873],
        [ 0.3873,  0.3873,  0.3873,  0.3873,  0.3873],
        [ 1.1619,  1.1619,  1.1619,  1.1619,  1.1619]])</pre>

## Processing the Output of a Classifier
A classifier will produce a list of probabilities across each class.  Since we typically process more than one input at a time (batch training), we will be getting batch number of outputs.  Let's say our batch size for this example is 16 and the classifier is attempting to predict 1 of 5 different classes.

In [14]:
y_out_logits = torch.randn((16,5))
y_out_logits

tensor([[-1.2113,  0.6304, -1.4713, -1.3352, -0.4897],
        [ 0.1317,  0.3295,  0.3264, -0.4806,  1.1032],
        [ 2.5485,  0.3006, -0.5432, -1.0841,  1.4612],
        [-1.6279, -1.4801, -1.0631,  0.3630,  0.3995],
        [ 0.1457, -0.7345, -0.9873,  1.8512, -1.3437],
        [ 0.8535,  0.8811, -0.6522,  0.5810,  0.3561],
        [ 0.0160,  0.4019,  1.9538, -0.4460,  1.7102],
        [ 0.8944, -0.5458, -0.6418, -2.0526,  0.3467],
        [-0.6969, -0.0047, -0.3136, -1.2602,  0.6977],
        [ 0.3720, -0.2606, -0.7613,  0.1277,  0.1522],
        [-1.1083, -0.6452, -1.7871,  0.6950, -0.5825],
        [-0.1926,  1.2357, -0.8008, -0.2808,  0.8701],
        [-1.7344, -1.4347, -0.0628, -0.5595,  1.0410],
        [ 0.1365,  1.8125, -0.4949,  0.8339, -0.3968],
        [ 1.3933,  0.3012, -0.2570, -1.3999,  0.7615],
        [ 0.8896, -0.3903,  0.8299,  0.2927, -0.6837]])

### Correct Answer:  
<pre>
tensor([[-1.2113,  0.6304, -1.4713, -1.3352, -0.4897],
        [ 0.1317,  0.3295,  0.3264, -0.4806,  1.1032],
        [ 2.5485,  0.3006, -0.5432, -1.0841,  1.4612],
        [-1.6279, -1.4801, -1.0631,  0.3630,  0.3995],
        [ 0.1457, -0.7345, -0.9873,  1.8512, -1.3437],
        [ 0.8535,  0.8811, -0.6522,  0.5810,  0.3561],
        [ 0.0160,  0.4019,  1.9538, -0.4460,  1.7102],
        [ 0.8944, -0.5458, -0.6418, -2.0526,  0.3467],
        [-0.6969, -0.0047, -0.3136, -1.2602,  0.6977],
        [ 0.3720, -0.2606, -0.7613,  0.1277,  0.1522],
        [-1.1083, -0.6452, -1.7871,  0.6950, -0.5825],
        [-0.1926,  1.2357, -0.8008, -0.2808,  0.8701],
        [-1.7344, -1.4347, -0.0628, -0.5595,  1.0410],
        [ 0.1365,  1.8125, -0.4949,  0.8339, -0.3968],
        [ 1.3933,  0.3012, -0.2570, -1.3999,  0.7615],
        [ 0.8896, -0.3903,  0.8299,  0.2927, -0.6837]])</pre>

In [15]:
# Check the size
y_out_logits.size()

torch.Size([16, 5])

### Correct Answer:  
<pre>
torch.Size([16, 5])</pre>

In [16]:
# Convert output logits to probabilities
y_outprob = torch.nn.functional.softmax(y_out_logits, dim=1)
y_outprob

tensor([[0.0907, 0.5724, 0.0700, 0.0802, 0.1867],
        [0.1511, 0.1842, 0.1836, 0.0819, 0.3992],
        [0.6602, 0.0697, 0.0300, 0.0175, 0.2226],
        [0.0531, 0.0616, 0.0934, 0.3888, 0.4032],
        [0.1339, 0.0555, 0.0431, 0.7372, 0.0302],
        [0.2763, 0.2840, 0.0613, 0.2104, 0.1680],
        [0.0646, 0.0950, 0.4484, 0.0407, 0.3514],
        [0.4801, 0.1137, 0.1033, 0.0252, 0.2776],
        [0.1103, 0.2204, 0.1618, 0.0628, 0.4448],
        [0.2908, 0.1545, 0.0936, 0.2277, 0.2334],
        [0.0921, 0.1463, 0.0467, 0.5590, 0.1558],
        [0.1050, 0.4379, 0.0571, 0.0961, 0.3038],
        [0.0371, 0.0501, 0.1974, 0.1201, 0.5953],
        [0.1056, 0.5642, 0.0562, 0.2121, 0.0619],
        [0.4716, 0.1582, 0.0906, 0.0289, 0.2507],
        [0.3358, 0.0934, 0.3163, 0.1849, 0.0696]])

### Correct Answer:  
<pre>
tensor([[0.0907, 0.5724, 0.0700, 0.0802, 0.1867],
        [0.1511, 0.1842, 0.1836, 0.0819, 0.3992],
        [0.6602, 0.0697, 0.0300, 0.0175, 0.2226],
        [0.0531, 0.0616, 0.0934, 0.3888, 0.4032],
        [0.1339, 0.0555, 0.0431, 0.7372, 0.0302],
        [0.2763, 0.2840, 0.0613, 0.2104, 0.1680],
        [0.0646, 0.0950, 0.4484, 0.0407, 0.3514],
        [0.4801, 0.1137, 0.1033, 0.0252, 0.2776],
        [0.1103, 0.2204, 0.1618, 0.0628, 0.4448],
        [0.2908, 0.1545, 0.0936, 0.2277, 0.2334],
        [0.0921, 0.1463, 0.0467, 0.5590, 0.1558],
        [0.1050, 0.4379, 0.0571, 0.0961, 0.3038],
        [0.0371, 0.0501, 0.1974, 0.1201, 0.5953],
        [0.1056, 0.5642, 0.0562, 0.2121, 0.0619],
        [0.4716, 0.1582, 0.0906, 0.0289, 0.2507],
        [0.3358, 0.0934, 0.3163, 0.1849, 0.0696]])</pre>

In [17]:
# Check that each sample sums up to 1
y_outprob.sum(dim=1)

tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
        1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000])

### Correct Answer:  
<pre>
tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
        1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000])</pre>

In [18]:
# Find the class predicted for each of the 16 inputs
prob, idx = torch.topk(y_outprob,1)
print (f"Probability associated with each predicted class:\n{prob}")
print (f"Class ID:\n{idx}")

Probability associated with each predicted class:
tensor([[0.5724],
        [0.3992],
        [0.6602],
        [0.4032],
        [0.7372],
        [0.2840],
        [0.4484],
        [0.4801],
        [0.4448],
        [0.2908],
        [0.5590],
        [0.4379],
        [0.5953],
        [0.5642],
        [0.4716],
        [0.3358]])
Class ID:
tensor([[1],
        [4],
        [0],
        [4],
        [3],
        [1],
        [2],
        [0],
        [4],
        [0],
        [3],
        [1],
        [4],
        [1],
        [0],
        [0]])


### Correct Answer:  
<pre>
Probability associated with each predicted class:
tensor([[0.5724],
        [0.3992],
        [0.6602],
        [0.4032],
        [0.7372],
        [0.2840],
        [0.4484],
        [0.4801],
        [0.4448],
        [0.2908],
        [0.5590],
        [0.4379],
        [0.5953],
        [0.5642],
        [0.4716],
        [0.3358]])
Class ID:
tensor([[1],
        [4],
        [0],
        [4],
        [3],
        [1],
        [2],
        [0],
        [4],
        [0],
        [3],
        [1],
        [4],
        [1],
        [0],
        [0]])</pre>

## Move Tensor to CPU or GPU
A tensor will either reside on a GPU or CPU depending on hardware acceleration mode selected.  With Colab, if you run the notebook using hardware acceleration (GPU), you will be able to place the tensor on a device("GPU").

In [19]:
# Determine environment
device = 'cuda' if torch.cuda.is_available() else 'cpu'
print (f"Device: {device}")

Device: cuda


### Correct Answer:  
<pre>
Device: cpu
</pre>
Note:  If you are using hardware acceleration, your device would be set to "cuda"

In [20]:
# Move the tensor to GPU if availiable
a=a.to(device)

## Experiment with Tensor Shapes
In some cases, you will need to add a batch dimension or remove an unused dimension from a tensor.  We will explore both below.

In [21]:
# Create a single dimension tensor with 10 values from 0..9
c = torch.arange(0,10,dtype=torch.float)
print (c)
print (c.size())

tensor([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])
torch.Size([10])


### Correct Answer:  
<pre>
tensor([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])
torch.Size([10])
</pre>

In [22]:
# Resize the tensor to 2 x 5
c = c.reshape(2,5)
print (c)
print (c.size())

tensor([[0., 1., 2., 3., 4.],
        [5., 6., 7., 8., 9.]])
torch.Size([2, 5])


### Correct Answer:  
<pre>
tensor([[0., 1., 2., 3., 4.],
        [5., 6., 7., 8., 9.]])
torch.Size([2, 5])
</pre>

In [23]:
# Reshape the tensor to 5 x 2
c = c.reshape(5,2)
print (c)
print (c.size())

tensor([[0., 1.],
        [2., 3.],
        [4., 5.],
        [6., 7.],
        [8., 9.]])
torch.Size([5, 2])


### Correct Answer:  
<pre>
tensor([[0., 1.],
        [2., 3.],
        [4., 5.],
        [6., 7.],
        [8., 9.]])
torch.Size([5, 2])
</pre>

In [24]:
# Add a single dimension (size [5, 2, 1])
c = c.reshape(5,2,1)
print (c)
print (c.size())

tensor([[[0.],
         [1.]],

        [[2.],
         [3.]],

        [[4.],
         [5.]],

        [[6.],
         [7.]],

        [[8.],
         [9.]]])
torch.Size([5, 2, 1])


### Correct Answer:  
<pre>
tensor([[[0.],
         [1.]],

        [[2.],
         [3.]],

        [[4.],
         [5.]],

        [[6.],
         [7.]],

        [[8.],
         [9.]]])
torch.Size([5, 2, 1])
</pre>

In [25]:
# Restore the shape by removing the single dimension components
c = c.squeeze()
print (c)
print (c.size())

tensor([[0., 1.],
        [2., 3.],
        [4., 5.],
        [6., 7.],
        [8., 9.]])
torch.Size([5, 2])


### Correct Answer:  
<pre>
tensor([[0., 1.],
        [2., 3.],
        [4., 5.],
        [6., 7.],
        [8., 9.]])
torch.Size([5, 2])
</pre>

In [26]:
# Flatten the tensor (size [10])
c = c.flatten()
print (c)
print (c.size())

tensor([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])
torch.Size([10])


### Correct Answer:  
<pre>
tensor([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])
torch.Size([10])
</pre>

In [27]:
# Add a batch dimension (ie, shape 1,10)
c = c.reshape(1,10)
print (c)
print (c.size())

tensor([[0., 1., 2., 3., 4., 5., 6., 7., 8., 9.]])
torch.Size([1, 10])


### Correct Answer:  
<pre>
tensor([[0., 1., 2., 3., 4., 5., 6., 7., 8., 9.]])
torch.Size([1, 10])
</pre>

### Key Takeaways
Key Takeaways:
- You created multiple dimensional tensors
- You explored their properties using size, type and python-style indexing/slicing
- You normalized a tabular set of data by calculating the mean and standard deviation for each column
- You converted a simulated set of logits to probabilities
- You used the topk function to determine the class ID predicted by the classifier for all 16 inputs
- You experimented with reshaping tensors